File: java/src/feb/week1/boj/스타트택시_박재환.java
package feb.week1.boj;

import java.util.*;
import java.io.*;

public class 스타트택시_박재환 {
    static BufferedReader br;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        init();
        br.close();
    }
    /**
     * 손님을 도착지로 데려다줄 때마다 연료가 충전된다. -> 승객을 태워 이동하면서 소모한 연료양의 두 배 만큼
     * 연료가 바닥나면 업무가 끝난다.
     *
     * M명의 승객을 태우는 것이 목표
     * N x N 격자
     * - 비어있음
     * - 벽이 있음
     *
     * 택시는 상하좌우 인접한 빈 칸 하나로 이동 가능
     * -> 늘 최단경로로만 움직임
     *
     * => 목적지 도착과, 연료 바닥이 동시에 나면, 이는 실패로 간주하지 않음
     */
    static StringTokenizer st;
    static int n, m;
    static int[][] map;
    static Taxi taxi;
    static Customer[] customers;
    static void init() throws IOException {
        st = new StringTokenizer(br.readLine().trim());
        n = Integer.parseInt(st.nextToken());       // 격자 크기
        m = Integer.parseInt(st.nextToken());       // 목표 손님 수
        int fuel = Integer.parseInt(st.nextToken());    // 초기 연료
        map = new int[n][n];
        for(int x=0; x<n; x++) {
            st = new StringTokenizer(br.readLine().trim());
            for(int y=0; y<n; y++) {
                map[x][y] = Integer.parseInt(st.nextToken());
                if(map[x][y] == 1) map[x][y] = -1;     // 벽을 -1 로 표시
            }
        }

        st = new StringTokenizer(br.readLine().trim());
        int tx = Integer.parseInt(st.nextToken())-1;
        int ty = Integer.parseInt(st.nextToken())-1;
        taxi = new Taxi(tx, ty, fuel);

        customers = new Customer[m];
        for(int i=0; i<m; i++) {
            st = new StringTokenizer(br.readLine().trim());
            int sx = Integer.parseInt(st.nextToken())-1;
            int sy = Integer.parseInt(st.nextToken())-1;
            int ex = Integer.parseInt(st.nextToken())-1;
            int ey = Integer.parseInt(st.nextToken())-1;
            map[sx][sy] = i+1;      // 손님 id 값 저장
            customers[i] = new Customer(sx, sy, ex, ey);
        }

        int result = solution();
        System.out.println(result);
    }
    static class Customer {
        int sx, sy;
        int ex, ey;
        int distFromTaxi;
        boolean done;

        Customer(int sx, int sy, int ex, int ey) {
            this.sx = sx;
            this.sy = sy;
            this.ex = ex;
            this.ey = ey;
            this.done = false;
        }
    }
    static class Taxi {
        int x, y;
        int fuel;
        int customerCount;

        Taxi(int x, int y, int fuel) {
            this.x = x;
            this.y = y;
            this.fuel = fuel;
            this.customerCount = 0;
        }
    }
    //--------------------------------------
    static int solution() {
        /**
         * 매번 손님을 태울 수 있는지 확인한다.
         */
        while(true) {
            int customerId = findNearCustomer();
            if(customerId == -1) break;
            // 손님을 찾은경우
            // 태울 수 있는지 확인
            Customer customer = customers[customerId];
            if(taxi.fuel < customer.distFromTaxi) break;       // 손님을 태우러 이동할 수 없음
            taxi.fuel -= customer.distFromTaxi;

            map[customer.sx][customer.sy] = 0;                  // 손님을 태웠다는 처리
            customer.done = true;
            taxi.x = customer.sx;
            taxi.y = customer.sy;
            // 목적지까지 이동할 수 있는지 확인
            int distToEnd = getDistToEnd(customer);
            if(distToEnd == -1) break;      // 목적지까지 이동할 수 있는 경로가 없음
            if(taxi.fuel < distToEnd) break;        // 연료가 부족함

            taxi.fuel -= distToEnd;
            taxi.fuel += distToEnd * 2;
            taxi.customerCount++;
            taxi.x = customer.ex;
            taxi.y = customer.ey;
        }

        return taxi.customerCount == m ? taxi.fuel : -1;
    }
    static int[] dx = {0,1,0,-1};
    static int[] dy = {1,0,-1,0};
    static int findNearCustomer() {
        /**
         * 같은 거리에 여러 승객이 있을 경우, 행 번호가 가장 작은 승객을
         * 행 번호가 같다면 열 번호가 가장 작은 승객을 고른다.
         */
        Queue<int[]> q = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][n];
        int[] foundCustomer = new int[3];
        foundCustomer[0] = n+1;
        foundCustomer[1] = n+1;
        foundCustomer[2] = n*n+1;
        // 만약 택시 위치에 손님이 있다면
        if(map[taxi.x][taxi.y] > 0) {
            int customerId = map[taxi.x][taxi.y]-1;
            customers[customerId].distFromTaxi = 0;
            return customerId;
        }
        // 택시위치 초기화
        q.offer(new int[] {taxi.x, taxi.y, 0});
        visited[taxi.x][taxi.y] = true;

        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            int dist = cur[2];
            if(dist > foundCustomer[2]) break;
            for(int dir=0; dir<4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];
                if(nx < 0 || ny < 0 || nx >= n || ny >= n) continue;
                if(visited[nx][ny] || map[nx][ny] == -1) continue;
                // 현재 위치에 손님이 있는지 확인
                if(map[nx][ny] > 0) {
                    if(foundCustomer[0] == n+1 && foundCustomer[1] == n+1 && foundCustomer[2] == n*n+1) {
                        foundCustomer[0] = nx;
                        foundCustomer[1] = ny;
                        foundCustomer[2] = dist+1;
                    } else {
                        if(foundCustomer[2] > dist+1) {
                            foundCustomer[0] = nx;
                            foundCustomer[1] = ny;
                            foundCustomer[2] = dist+1;
                        } else if(foundCustomer[2] == dist+1 && foundCustomer[0] > nx) {
                            foundCustomer[0] = nx;
                            foundCustomer[1] = ny;
                        } else if(foundCustomer[2] == dist+1 && foundCustomer[0] == nx && foundCustomer[1] > ny) {
                            foundCustomer[1] = ny;
                        }
                    }
                }
                visited[nx][ny] = true;
                q.offer(new int[] {nx, ny, dist+1});
            }
        }

        if(foundCustomer[0] == n+1 && foundCustomer[1] == n+1 && foundCustomer[2] == n*n+1) return -1;
        int customerId = map[foundCustomer[0]][foundCustomer[1]]-1;
        customers[customerId].distFromTaxi = foundCustomer[2];
        return customerId;
    }
    static int getDistToEnd(Customer customer) {
        Queue<int[]> q = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][n];

        int sx = customer.sx, sy = customer.sy;
        int ex = customer.ex, ey = customer.ey;
        q.offer(new int[] {sx, sy, 0});
        visited[sx][sy] = true;

        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            int dist = cur[2];

            if(x == ex && y == ey) return dist;

            for(int dir=0; dir<4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];
                if(nx < 0 || ny < 0 || nx >= n || ny >= n) continue;
                if(visited[nx][ny] || map[nx][ny] == -1) continue;
                visited[nx][ny] = true;
                q.offer(new int[] {nx, ny, dist+1});
            }
        }

        return -1;
    }
}
/*
6 3 15
0 0 1 0 0 0
0 0 1 0 0 0
0 0 0 0 0 0
0 0 0 0 0 0
0 0 0 0 1 0
0 0 0 1 0 0
6 5
2 2 5 6
5 4 1 6
4 2 6 5

14
 */

File: python/src/feb/week1/boj/스타트택시_박재환.py
def solution():
    import sys
    input = sys.stdin.readline

    n, m, fuel = map(int, input().split())
    board = list(list(map(int, input().split())) for _ in range(n))
    for x in range(n):
        for y in range(n):
            if board[x][y] == 1:
                board[x][y] = -1

    tx, ty = map(int, input().split())
    taxi = [tx-1, ty-1, fuel, 0]        # x, y, 남은 연료, 태운 손님

    customers = []
    for i in range(m):
        sx, sy, ex, ey = map(int, input().split())
        customers.append([sx-1, sy-1, ex-1, ey-1])
        board[sx-1][sy-1] = i+1

    dx = [-1, 0, 1, 0]
    dy = [0, 1, 0, -1]

    def find_near_customer():
        from collections import deque
        visited = [[False] * n for _ in range(n)]
        q = deque()

        # 시작 위치
        q.append((taxi[0], taxi[1]))
        visited[taxi[0]][taxi[1]] = True

        dist = 0
        while q:
            size = len(q)
            candidates = []

            for _ in range(size):
                x, y = q.popleft()

                if board[x][y] > 0:
                    candidates.append((x, y))

                for d in range(4):
                    nx = x + dx[d]
                    ny = y + dy[d]

                    if nx < 0 or ny < 0 or nx >= n or ny >= n:
                        continue
                    if visited[nx][ny] or board[nx][ny] == -1:
                        continue

                    visited[nx][ny] = True
                    q.append((nx, ny))

            if candidates:
                candidates.sort()  # 행 → 열 우선
                cx, cy = candidates[0]
                return board[cx][cy] - 1, dist

            dist += 1

        return -1, -1
    def get_dist_to_end(customer_id):
        from collections import deque
        q = deque()
        visited = [[False] * n for _ in range(n)]

        sx, sy, ex, ey = customers[customer_id]
        q.append((sx, sy, 0))
        visited[sx][sy] = True

        while q:
            x, y, dist = q.popleft()
            if x == ex and y == ey:
                return dist
            for dir in range(4):
                nx = x + dx[dir]
                ny = y + dy[dir]

                if nx < 0 or ny < 0 or nx >= n or ny >= n:
                    continue
                if visited[nx][ny] or board[nx][ny] == -1:
                    continue
                visited[nx][ny] = True
                q.append((nx, ny, dist + 1))
        return -1

    while True:
        customer_id, c_dist = find_near_customer()
        if customer_id == -1:
            break

        # 손님을 찾았다면
        # 손님의 출발지까지 이동가능한지
        if taxi[2] < c_dist:
            break
        taxi[2] -= c_dist
        taxi[0] = customers[customer_id][0]
        taxi[1] = customers[customer_id][1]
        board[customers[customer_id][0]][customers[customer_id][1]] = 0

        dist_to_end = get_dist_to_end(customer_id)
        if dist_to_end == -1 or dist_to_end > taxi[2]:
            break
        taxi[2] -= dist_to_end
        taxi[0] = customers[customer_id][2]
        taxi[1] = customers[customer_id][3]
        taxi[2] += dist_to_end * 2
        taxi[3] += 1

    return taxi[2] if taxi[3] == m else -1

if __name__ == '__main__':
    print(solution())